docs/man/man/NetworkController.3
.TH "NetworkController" 3 "Mon Jan 5 2026" "Version 1.0" "AmurCore" \" -*- nroff -*-
.ad l
.nh
.SH NAME
NetworkController
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <networkcontroller\&.h>\fP
.PP
Inherits QObject\&.
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBsensorsUpdated\fP ()"
.br
.ti -1c
.RI "void \fBmapUpdated\fP ()"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBNetworkController\fP (std::shared_ptr< Controls > controlsPtr, std::shared_ptr< Sensors > sensorsPtr, std::shared_ptr< map_service::GetMapResponse > mapPtr)"
.br
.ti -1c
.RI "\fB~NetworkController\fP ()"
.br
.ti -1c
.RI "int \fBrunClient\fP (std::string &server_address)"
.br
.ti -1c
.RI "int \fBrunServer\fP (std::string &address_mask)"
.br
.ti -1c
.RI "int \fBrunArpingService\fP (int arpPort, int gRPCPort, QString &arpHeader)"
.br
.ti -1c
.RI "QList< std::shared_ptr< \fBRobotEntry\fP > > \fBgetRobots\fP () const"
.br
.ti -1c
.RI "std::shared_ptr< \fBgrpcClient\fP > & \fBgetClientInstance\fP ()"
.br
.ti -1c
.RI "std::shared_ptr< \fBgrpcServer\fP > & \fBgetServerInstance\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "std::shared_ptr< Controls > \fBcontrols\fP"
.br
.ti -1c
.RI "std::shared_ptr< Sensors > \fBsensors\fP"
.br
.ti -1c
.RI "std::shared_ptr< map_service::GetMapResponse > \fBmap\fP"
.br
.ti -1c
.RI "grpc::Status \fBclientStatus\fP"
.br
.ti -1c
.RI "grpc::ServerBuilder \fBbuilder\fP"
.br
.ti -1c
.RI "std::shared_ptr< \fBgrpcServer\fP > \fBserverPtr\fP"
.br
.ti -1c
.RI "std::shared_ptr< \fBgrpcClient\fP > \fBclientPtr\fP"
.br
.ti -1c
.RI "int \fBmessageNo\fP = 1"
.br
.ti -1c
.RI "QUdpSocket * \fBudpSocket\fP = nullptr"
.br
.ti -1c
.RI "QList< std::shared_ptr< \fBRobotEntry\fP > > \fBrobots\fP"
.br
.ti -1c
.RI "QMap< quint16, QUdpSocket * > \fBclientSockets\fP"
.br
.in -1c
.SH "Constructor & Destructor Documentation"
.PP 
.SS "NetworkController::NetworkController (std::shared_ptr< Controls > controlsPtr, std::shared_ptr< Sensors > sensorsPtr, std::shared_ptr< map_service::GetMapResponse > mapPtr)"

.SS "NetworkController::~NetworkController ()"

.SH "Member Function Documentation"
.PP 
.SS "std::shared_ptr<\fBgrpcClient\fP>& NetworkController::getClientInstance ()\fC [inline]\fP"

.SS "QList< std::shared_ptr< \fBRobotEntry\fP > > NetworkController::getRobots () const"

.SS "std::shared_ptr<\fBgrpcServer\fP>& NetworkController::getServerInstance ()\fC [inline]\fP"

.SS "void NetworkController::mapUpdated ()\fC [signal]\fP"

.SS "int NetworkController::runArpingService (int arpPort, int gRPCPort, QString & arpHeader)"

.SS "int NetworkController::runClient (std::string & server_address)"

.SS "int NetworkController::runServer (std::string & address_mask)"

.SS "void NetworkController::sensorsUpdated ()\fC [signal]\fP"

.SH "Member Data Documentation"
.PP 
.SS "grpc::ServerBuilder NetworkController::builder\fC [private]\fP"

.SS "std::shared_ptr<\fBgrpcClient\fP> NetworkController::clientPtr\fC [private]\fP"

.SS "QMap<quint16, QUdpSocket*> NetworkController::clientSockets\fC [private]\fP"

.SS "grpc::Status NetworkController::clientStatus\fC [private]\fP"

.SS "std::shared_ptr<Controls> NetworkController::controls\fC [private]\fP"

.SS "std::shared_ptr<map_service::GetMapResponse> NetworkController::map\fC [private]\fP"

.SS "int NetworkController::messageNo = 1\fC [private]\fP"

.SS "QList< std::shared_ptr<\fBRobotEntry\fP> > NetworkController::robots\fC [private]\fP"

.SS "std::shared_ptr<Sensors> NetworkController::sensors\fC [private]\fP"

.SS "std::shared_ptr<\fBgrpcServer\fP> NetworkController::serverPtr\fC [private]\fP"

.SS "QUdpSocket* NetworkController::udpSocket = nullptr\fC [private]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for AmurCore from the source code\&.


CHANGELOG.md
## 1.0.0 (2025-02-28)

### BREAKING CHANGE

- changed proto file
- Many changes at NetworkController

### Feat

- version with follow waypoints & show pose and map
- **GUI**: added logo, removed speech from menu & added map, robot info entries
- added dialog & widget for showing global map & robot. added robot info dialog
- **protos**: splitted proto files & added global map recieving with robot pose
- added lookup table for robots with storing in database, grpc build related fixes & added config controller

### Fix

- changed raw pointers to shared_ptr
- **speech**: removed unused speech processor

## 0.2.0 (2023-04-30)

### Feat

- **NetworkController**: Add finding robots by broadcast request
- **NetworkController**: Added first arp message via JSON

### Fix

- **CMakeLists**: Fix some problems with compiling SpeechProccessor
- **NetworkController**: Removed forgotten include to nonexisted file
- **AmurCore.pro**: Fix gRPC linking error & add commitizen config
- **CamCalibrate**: Fixed waiting new frame with chessboard, changed to std::chrono lib
- **JSONWorker**: Disable JSONWorker, for possible building


network/networkcontroller.cpp
#include "networkcontroller.h"

NetworkController::NetworkController(std::shared_ptr<Controls> controlsPtr,
                                     std::shared_ptr<Sensors> sensorsPtr,
                                     std::shared_ptr<map_service::GetMapResponse> mapPtr)  // TODO - add robot id & &<vector> of robots id
    : controls(controlsPtr),
      sensors(sensorsPtr),
      map(mapPtr)
{
    udpSocket = new QUdpSocket(this);
    serverPtr = NetworkFactory::createServer();
}

NetworkController::~NetworkController()
{
    // Clear list of robots
    robots.clear();
}

int NetworkController::runClient(std::string &server_address) // TODO - send const &<vector> of robots id
{
    // Instantiate the client. It requires a channel, out of which the actual RPCs
    // are created. This channel models a connection to an endpoint specified by
    // the argument "--target=" which is the only expected argument.
    // We indicate that the channel isn't authenticated (use of
    // InsecureChannelCredentials()).

    clientPtr = NetworkFactory::createClient(server_address, controls, sensors, map);

    QObject::connect(clientPtr.get(), &grpcClient::sensorsUpdated, this, &NetworkController::sensorsUpdated);
    QObject::connect(clientPtr.get(), &grpcClient::mapUpdated, this, &NetworkController::mapUpdated);

    std::thread thr([&, this]()
     {

        clientStatus = this->clientPtr->DataStreamExchange();
        std::cout << "State is OK?: " << clientStatus.ok() << std::endl;
     }
    );
    thr.detach();

    return 0;
}

int NetworkController::runServer(std::string &address_mask) // TODO - send const &<vector> of robots id
{
    std::thread thr([&]()
     {
      grpc::EnableDefaultHealthCheckService(true);
      grpc::reflection::InitProtoReflectionServerBuilderPlugin();

      // Send protos pointers to server
      serverPtr->setProtosPointers(controls, sensors, map);
      serverPtr->setSensorsUpdatedCallback([this]() { emit sensorsUpdated(); });

      // Listen on the given address without any authentication mechanism.
      builder.AddListeningPort(address_mask, grpc::InsecureServerCredentials());

      // Register "service" as the instance through which we'll communicate with
      // clients. In this case it corresponds to an *synchronous* service.
      builder.RegisterService(serverPtr.get());

      // Finally assemble the server.
      std::unique_ptr<grpc::Server> server(builder.BuildAndStart());
      std::cout << "Server listening on " << address_mask << std::endl;

      // Wait for the server to shutdown. Note that some other thread must be
      // responsible for shutting down the server for this call to ever return.
      server->Wait();

      std::cout << "Server stopped " << std::endl;
     }
    );
    thr.detach();

    return 0;
}

int NetworkController::runArpingService(int arpPort, int gRPCPort, QString &arpHeader)
{
    udpSocket->bind(QHostAddress::AnyIPv4, arpPort);

    QObject::connect(udpSocket, &QUdpSocket::readyRead, [&, gRPCPort](){
        while (udpSocket->hasPendingDatagrams()) {
            QByteArray datagram;
            datagram.resize(udpSocket->pendingDatagramSize());

            QHostAddress senderAddress;
            quint16 senderBCastPort;
            udpSocket->readDatagram(datagram.data(), datagram.size(), &senderAddress, &senderBCastPort);

            // Split message by separator
            QString message = QString::fromUtf8(datagram);
            QStringList parsed = message.split(u':');

            // Process message if it have preambula - "AMUR:"
            if(parsed.empty())
                continue;

            if(arpHeader == parsed.at(0)){
                qDebug() << "Received arping message: " << message << " from " << senderAddress.toString() << ":" << senderBCastPort;

                QString machineID;
                if(parsed.size() > 1)
                    machineID = parsed.at(1);

                int arpingPort = 0;
                if(parsed.size() > 2)
                    arpingPort = parsed.at(2).toInt();

                // Check if the client is already in the list
                std::shared_ptr<RobotEntry> robot;
                for (int i = 0; i < robots.size(); ++i) {
                    if (robots[i]->address() == senderAddress) {
                        robot = robots[i];
                        break;
                    }
                }

                // Add a new client if it not found in list of robots
                if (!robot) {
                    robot = std::make_shared<RobotEntry>(udpSocket, senderAddress, senderBCastPort);
                    robot->setMachineID(machineID);

                    // set custom port for answer if setted
                    if(arpingPort > 0)
                        robot->setPort(arpingPort);

                    robots.append(robot);
                    qDebug() << "Added new client " << senderAddress.toString() << ":" << arpingPort;
                }

                QByteArray response = "OK:" + QByteArray::number(gRPCPort);
                robot->sendData(response);
            }

        }
    });

    std::cout << "" << std::endl;
    return 0;
}

QList< std::shared_ptr<RobotEntry> > NetworkController::getRobots() const
{
    return robots;
}
